### PR TITLE
Discovery local

### PR DESCRIPTION
**Pull request by herzrasen**
This is a simple implementation for #157. Using the `akka.discovery.method = akka-local` uses `akka.discovery.akka-local.service-file` as a register for locally running services. When shutting down the service the shutdownHook will unregister the stopping service. 

Input and suggestions are welcome.

### CODE REVIEW

File: discovery-local/src/main/scala/akka/discovery/local/registration/LocalServiceRegistration.scala [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+  /**
+   * Removes an <host>:<port> from the service file if registered.
+   * @param address the hostname or ip address
+   * @param port the port
+   */
+  def remove(address: String, port: Int): Unit =
+    writeToFile(localServiceEntries.filterNot(_ == LocalServiceEntry(address, port)))
+
+  private def writeToFile(localServiceEntries: Seq[LocalServiceEntry]) =
+    Files.write(serviceFile, localServiceEntries.toJson.compactPrint.getBytes(utf8))
```
Comment (ktoso) [inline discovery-local/src/main/scala/akka/discovery/local/registration/LocalServiceRegistration.scala]: You could get 2 discoveries attempting to write and corrupt the file this way. Needs to take filesystem lock https://docs.oracle.com/javase/7/docs/api/java/nio/channels/FileLock.html around these operations. This likely is good enough as we're not talking about any high performance stuff here, just avoiding writing into it from more apps

File: discovery-local/src/main/scala/akka/discovery/local/LocalServiceDiscovery.scala [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+    log.debug(s"Using $serviceFile as service file")
+    val addresses =
+      localServiceRegistration.localServiceEntries.map { entry =>
+        ResolvedTarget(entry.addr, Some(entry.port))
+      }
+
+    // remove self from the service file on shutdown
+    sys.addShutdownHook {
+      unregisterSelf()
+    }
```
Comment (ktoso) [inline discovery-local/src/main/scala/akka/discovery/local/LocalServiceDiscovery.scala]: this should rather use the coordinated shutdown hooks, not a system wide one.
See https://doc.akka.io/docs/akka/current/actors.html#coordinated-shutdown

**Comment by herzrasen** [comment]
@ktoso thanks for your remarks. I updated the sources accordingly.

**Comment by herzrasen** [comment]
Any chance this is going to be merged?

**Comment by chbatey** [comment]
It is very similar to the config based service discovery. However maybe there's a usecase for file based that is not inside the application's config? WDYT @herzrasen @ktoso ?

**Comment by herzrasen** [comment]
@chbatey it is indeed very similar. I use it in testing environments where no DNS is present. I could basically do the same with an SEED_NODES environement variable which is only set in my test envs.

**Comment by ktoso** [comment]
I was somewhat in progress of reviewing this, but the general feel is that this is not quite what i had envisioned as the one I had in mind was also about registering into it on startup becoming a "super easy no need to configure anything mode" for demos and playing around. Not sure if we'd like to head there with this PR, WDYT?

**Comment by herzrasen** [comment]
I understand what you mean. I'm just not quite sure how this could be implemented since some kind of configuration is needed in any case.

**Pull request closed without merging**